enhancements/cloud-integration/tokenized-auth-enablement-operators-on-cloud.md

```diff
@@ -0,0 +1,503 @@
+under the Red Hat brand (that need to reach cloud APIs).
+ 
+As the managed services, including HyperShift teams, offering a downstream opinionated, supported and managed lifecycle 
+of OpenShift (in the forms of ROSA, ARO, OSD on GCP, HyperShift, etc), the OpenShift platform should have as close as 
+possible, native integration with core platform operators when clusters use tokenized cloud auth, driving the use of 
+layered products.
+
+As the Hypershift team, where the only credential mode for clusters/customers is STS (on AWS) , the Red Hat branded 
+Operators that must reach the AWS API, should be enabled to work with STS credentials in a consistent, and automated 
+fashion that allows customer to use those operators as easily as possible, driving the use of layered products.
+
+As an operator team that may already be supporting cloud credential provisioning, including a manner for allowing
+operand instances to have a unique set of credential per instance, I want my solution to keep working (Quay Operator).
+
+As a customer of OpenShift layered products I can confidently upgrade the operator even if the next version requires
+elevated permissions be provided.
+
+As an operator author I have a set of best practices and tools that help me develop a compliant operator.
+
+As a monitor of OpenShift layered products I can identify which operators are compliant with STS best practices and 
+tools. 
+
+As a cluster admin, it's clear when installing an operator if it is STS compliant (i.e. either doesn't need cloud 
+resources or does and is following the best practices).
+
+As an OpenShift user I want the STS enablement solution to be as cloud-agnostic as possible.
+
+### Goals
+
+Allow Day-2 Operators to access cloud provider resources as seamlessly as possible when being installed, used and
+updated on STS enabled clusters.
+
+While providing the above goal, allow for multi-tenancy. In this sense multi-tenancy means that an operator may
+provide different cloud provider credentials per operand such that operator users may uniquely access non-shared cloud
+resources.
+
+Operator authors have a way to notify, guide, and assist Day-2 Operator admins in providing the required cloud provider credentials matched to their permission
+needs for install and update.
+
+Ideally, a solution here will work in both HyperShift (STS always) and non-HyperShift clusters.
+
+### Non-Goals
+
+Day 1 operators are not included in this proposal, they are pre-provisioned for STS enabled clusters (at least 
+for HyperShift)
+
+
+
+## Proposal
+
+**Cloud Credential Operator (CCO) changes**: Add a "Token" mode, distinct from disabled/manual mode that will look for 
+and process CredentialRequests referenced in Operator CRs. This will operate in one of two ways depending on which other
+parts of this EP for other components are adopted: specifically HyperShift/Rosa adding CCO with Token mode or continuing
```

Comment:
As discussed, if CCO is deployed on Hypershift we would need to disable the controller that deploys the AWS pod identity webhook since Hypershift runs the pod identity webhook sidecar to the control plane.

The pod identity webhook deployment controller currently [checks the cluster infrastructure object](https://github.com/openshift/cloud-credential-operator/blob/bc2623480d290dcf9d4ea3ccc4717159f6799afd/pkg/operator/awspodidentity/awspodidentitywebhook_controller.go#L132-L139) to enable the controller when `.status.platform` is `AWS` and could also check for `.status.controlPlaneTopology: External` to determine that we're on a Hypershift cluster.

Reply:
> As discussed, if CCO is deployed on Hypershift we would need to disable the controller that deploys the AWS pod identity webhook

where was this discussed?  (can i read it?)

i'm wondering why it's a problem if both run(my understanding is the webhook looks for annotated SAs and injects a mount into the pod...  the CCO logic is just going to look for credreq objects and create a secret(not even force it to be mounted, though most likely pods will mount it)?  So it doesn't seem like the two are in direct conflict.

and also what are the implications may be on users/components that already expecting the AWS pod identity webhook to be running if it is disabled?

Reply:
> where was this discussed? (can i read it?)

This was brought up in one of our weekly sync meetings.

> i'm wondering why it's a problem if both run(my understanding is the webhook looks for annotated SAs and injects a mount into the pod... the CCO logic is just going to look for credreq objects and create a secret(not even force it to be mounted, though most likely pods will mount it)? So it doesn't seem like the two are in direct conflict.

> and also what are the implications may be on users/components that already expecting the AWS pod identity webhook to be running if it is disabled?

Hypershift is already running the pod identity webhook sidecar to the control plane pods (same pod identity webhook that CCO would deploy when installed on an AWS cluster) so just saying that we shouldn't deploy it because its already there.

Reply:
ah, sorry, i misunderstood.  per my other [comment](https://github.com/openshift/enhancements/pull/1339#discussion_r1170453839) my expectation is that this EP doesn't result in any situation where we deploy or rely on the AWS pod identity webhook.

Reply:
https://github.com/openshift/hypershift/blob/47ce832a0a0acc4690461676f3a00c4feb929a49/control-plane-operator/controllers/hostedcontrolplane/kas/deployment.go#L213-L244

Reply:
only as an alternative solution in the EP. Going to mark this resolved for now.